lib/app/lifecycle.coffee

```diff
@@ -119,16 +119,21 @@ crypto = require 'crypto'
   next err
 
 @ssoAndRedirectBack = (req, res, next) ->
-  return res.send { success: true, user: req.user.toJSON() } if req.xhr
-  parsed = parse redirectBack req
-  parsed = parse resolve opts.APP_URL, parsed.path unless parsed.hostname
-  domain = parsed.hostname?.split('.').slice(1).join('.')
-  return redirectBack(req, res) if domain isnt 'artsy.net'
+  urlToRedirectTo = parse redirectBack req
+  urlToRedirectTo = parse resolve opts.APP_URL, urlToRedirectTo.path unless urlToRedirectTo.hostname
+  domain = urlToRedirectTo.hostname?.split('.').slice(1).join('.')
+  # return redirectBack(req, res) if domain isnt 'artsy.net'
   request
     .post "#{opts.ARTSY_URL}/api/v1/me/trust_token"
     .set { 'X-Access-Token': req.user.get 'accessToken' }
     .end (err, sres) ->
-      return res.redirect parsed.href if err
-      res.redirect "#{opts.ARTSY_URL}/users/sign_in" +
+      return res.redirect urlToRedirectTo.href if err
+
+      apiAuthWithRedirectUrl = "#{opts.ARTSY_URL}/users/sign_in" +
         "?trust_token=#{sres.body.trust_token}" +
-        "&redirect_uri=#{parsed.href}"
+        "&redirect_uri=#{urlToRedirectTo.href}"
+
+      if req.xhr?
+        res.send { success: true, api_sign_in_with_redirect: apiAuthWithRedirectUrl, user: req.user.toJSON() }
```

Comment:
If I understand correctly, this gives clients the opportunity to navigate their user through the api and back to the current application. Isn't the client already able to do that though, since it specifies the redirect target in the first place?

Reply:
Can you elaborate? Not really following you here.

Reply:
I _think_ what @joeyAghion might be saying (correct me if I'm wrong!), is that, rather than encode this behavior in this lib (being able to provide the redirect location for a Gravity auth), maybe it can just stay in the client (Force)?

According to the [dependency graph](https://github.com/artsy/artsy-passport/network/dependents), it doesn't actually look like there are any other active clients besides Force at the moment. So if it's not realistic/feasible for _other_ clients to take advantage of this, it might be easier to just work in Force (or Reaction, if that's where authentication flow/modal stuff lives), and avoid faffing with this library. Those are more modern repos and technologies that are better supported (personally I view some of Artsy's older libs/repos, such as this one, as pseudo-deprecated, in that there's usually better options _today_, and these tend to rustle jimmies).

Reply:
> According to the dependency graph, it doesn't actually look like there are any other active clients besides Force at the moment. 

This library is being used by [artsy/vanity](https://github.com/artsy/vanity/blob/master/lib/setup.coffee#L11) as well.

>  So if it's not realistic/feasible for other clients to take advantage of this, it might be easier to just work in Force (or Reaction, if that's where authentication flow/modal stuff lives), and avoid faffing with this library. Those are more modern repos and technologies that are better supported (personally I view some of Artsy's older libs/repos, such as this one, as pseudo-deprecated, in that there's usually better options today, and these tend to rustle jimmies).

So a few things:

1. I'll acknowledge by bias here as someone that spent a significant amount of time learning up on cofeescript and related technologies here. I'm interested in solving this problem, and moving onto other problems :)

2. Is there a concrete plan to move off of artsy-passport? If so, I'm all for investing in solving this problem via a more supported solution path.

However, if there isn't, then I don't want **avoid** solution-ing within artsy-passport in this way because there's more modern technology other places. Moreover, there already seems to be a strong separation of concerns between "authentication" logic in artsy-passport and other view/route/business logic within force/reaction, and I think there's value in that code boundary.

---

If there are strong feelings on another concrete ways to solution here, I'm all ears. I'd appreciate a pair on it though!

I would like to goal set to have concrete next steps here by end of day (EST) tomorrow though.

cc/ @mzikherman @joeyAghion

Reply:
To clarify, yes, my question was simply: does artsy-passport _need_ to offer this redirect pattern (e.g., because otherwise some value such as an internal token wouldn't be accessible), or could a client decide to perform this redirect on its own?

This library is definitely in active use so I don't see any reason to _avoid_ updating it. I only asked in case we didn't _need_ to change this API for a particular client use case.

Reply:
In case these haven't been surfaced yet, there are some docs addressing redirects here: https://github.com/artsy/reaction/blob/c9d18e7f9ff47f2bf72f2f8f950d345d976d9d20/docs/authentication.md#redirects

Is the original form implementation using either `destination` or `redirectTo`, and we've confirmed it does not work? https://github.com/artsy/reaction/blob/20d28d5628c680270dfe01b4875a7732fd7a9576/src/Components/Authentication/Types.ts#L63

Reply:
> In case these haven't been surfaced yet, there are some docs addressing redirects here: https://github.com/artsy/reaction/blob/c9d18e7f9ff47f2bf72f2f8f950d345d976d9d20/docs/authentication.md#redirects

I've checked it out before, but thanks for the reminder!

> Is the original form implementation using either destination or redirectTo, and we've confirmed it does not work? https://github.com/artsy/reaction/blob/20d28d5628c680270dfe01b4875a7732fd7a9576/src/Components/Authentication/Types.ts#L63

Yup, confirmed it doesn't work, if by work, we mean create a session with Gravity. Check out [this GH comment](https://github.com/artsy/force/pull/4349#issuecomment-516882362) for more information.

WRT `redirectTo` or `destination` – my understanding of the code is that when calling [`ModalManager` in Force](https://github.com/artsy/force/blob/33ad0bf9f13c737a9e6993cf1bf9febc4f70c0b4/src/desktop/apps/authentication/components/ModalContainer.tsx#L75), we _always_ pass a `handleSubmit` prop, the value of which is the function defined in [`helpers.ts` which first considers the `redirectTo` of the modal options or respond with default logic](https://github.com/artsy/force/blob/79e11ee82200a990240f5a1b0e6a3fcd45dbba0e/src/desktop/apps/authentication/helpers.ts#L78).

To the best of my knowledge, there's only 1 place in force when `redirectTo` is set to a value that isn't the current window location, when registering for an auction.

<img width="444" alt="Screen Shot 2019-08-01 at 11 43 18 PM" src="https://user-images.githubusercontent.com/1561546/62342831-3f445080-b4b6-11e9-88cc-9899e88fefcf.png">

So one limitation here when logging in and registering at the same time for an auction, we won't be creating a session with Gravity.

---

To see how we could mitigate against the limitation mentioned above, I'm going to give implementing this logic within Force a stab by providing some form of wrapper around `afterSubmit`.